src/engine/source/router/src/runtimePolicy.cpp

```diff
@@ -32,14 +32,196 @@ std::optional<base::Error> RuntimePolicy::build(std::shared_ptr<builder::Builder
+                    if (innerIt == innerMap.end())
+                    {
+                        // There is no entry for the key in the second level, create a new one
+                        innerMap[key] = std::vector<std::shared_ptr<std::stringstream>> {traceStream};
+                    }
+                    else
+                    {
+                        // The key already exists at the second level, add the traceStream to the existing list
+                        innerIt->second.push_back(traceStream);
+                    }
+                }
+            }
+        }));
+}
+
+std::variant<std::tuple<std::string, std::string>, base::Error>
+RuntimePolicy::getData(const std::string& policyName, DebugMode debugMode, const std::string& assetTrace)
+{
+    std::unique_lock<std::shared_mutex> lock {m_mutexData};
+    auto trace = json::Json {R"({})"};
+    if (DebugMode::OUTPUT_AND_TRACES_WITH_DETAILS == debugMode)
+    {
+        if (m_history[policyName].empty())
+        {
+            m_traceBuffer[policyName].clear();
+            return base::Error {fmt::format(
+                "Policy '{}' has not been configured for trace tracking and output subscription", policyName)};
+        }
+
+        if (!assetTrace.empty())
+        {
+            for (const auto& [asset, condition] : m_history[policyName])
+            {
+                if (assetTrace == asset)
+                {
+                    if (m_traceBuffer.find(policyName) == m_traceBuffer.end())
+                    {
+                        m_traceBuffer[policyName].clear();
+                        return base::Error {fmt::format(
+                            "Policy '{}' has not been configured for trace tracking and output subscription",
+                            policyName)};
+                    }
+                    auto& tracePair = m_traceBuffer[policyName];
+                    if (tracePair.find(asset) != tracePair.end())
+                    {
+                        auto& traceVector = tracePair[asset];
+                        std::set<std::string> uniqueTraces; // Set for warehouses single traces
+                        for (const auto& traceStream : traceVector)
+                        {
+                            uniqueTraces.insert(traceStream->str()); // Insert unique traces in the set
+                        }
+                        std::stringstream combinedTrace;
+                        for (const auto& uniqueTrace : uniqueTraces)
+                        {
+                            combinedTrace << uniqueTrace;
+                        }
+                        trace.setString(combinedTrace.str(), std::string("/") + asset);
+                        m_traceBuffer[policyName].clear();
+                    }
+                }
+                else
+                {
+                    trace.setString(condition, std::string("/") + asset);
+                }
+            }
+        }
+        else
```

Comment:
I think we can make an effort of merging this 2 blocks in order to avoid code duplication, right?

Reply:
I agree